packages/block-library/src/template-part/edit/import-controls.js

```diff
@@ -0,0 +1,160 @@
+	const [ isBusy, setIsBusy ] = useState( false );
+
+	const registry = useRegistry();
+	const sidebars = useSelect( ( select ) => {
+		return select( coreStore ).getSidebars( {
+			per_page: -1,
+			_fields: 'id,name,description,status,widgets',
+		} );
+	}, [] );
+	const { getBlockTransformItems } = useSelect( blockEditorStore );
+	const { createErrorNotice } = useDispatch( noticesStore );
+
+	const createFromBlocks = useCreateTemplatePartFromBlocks(
+		area,
+		setAttributes
+	);
+
+	const options = useMemo( () => {
+		const sidebarOptions = ( sidebars ?? [] )
+			.filter( ( widgetArea ) => widgetArea.widgets.length > 0 )
+			.map( ( widgetArea ) => {
+				return {
+					value: widgetArea.id,
+					label: widgetArea.name,
+				};
+			} );
+
+		if ( ! sidebarOptions.length ) {
+			return [];
+		}
+
+		return [
+			{ value: '', label: __( 'Select widget area' ) },
+			...sidebarOptions,
+		];
+	}, [ sidebars ] );
+
+	async function createFromWidgets( event ) {
+		event.preventDefault();
+
+		if ( isBusy || ! selectedSidebar ) {
+			return;
+		}
+
+		setIsBusy( true );
+
+		const sidebar = options.find(
+			( { value } ) => value === selectedSidebar
+		);
+		const { getWidgets } = registry.resolveSelect( coreStore );
+
+		// The widgets API always returns a successful response.
+		const widgets = await getWidgets( {
+			sidebar: sidebar.value,
+			_embed: 'about',
+		} );
+
+		const skippedWidgets = new Set();
+		const blocks = widgets.flatMap( ( widget ) => {
+			const block = transformWidgetToBlock( widget );
+
+			if ( block.name !== 'core/legacy-widget' ) {
+				return block;
+			}
+
+			const transforms = getBlockTransformItems( block ).filter(
+				( item ) => ! WILDCARD_TRANSFORMS.includes( item.name )
+			);
+
+			// Skip the block if we have no matching transformations.
+			if ( ! transforms.length ) {
+				skippedWidgets.add( widget.id_base );
+				return [];
+			}
+
+			// Try transforming the Legacy Widget into a first matching block.
+			return switchToBlockType( block, transforms[ 0 ].name );
+		} );
+
+		await createFromBlocks(
+			blocks,
+			/* translators: %s: name of the widget area */
+			sprintf( __( 'Widget area: %s' ), sidebar.label )
+		);
+
+		if ( skippedWidgets.size ) {
+			createErrorNotice(
+				sprintf(
+					/* translators: %s: the list of widgets */
+					__( 'Unable to import the following widgets: %s.' ),
+					Array.from( skippedWidgets ).join( ', ' )
+				),
+				{
+					type: 'snackbar',
+				}
+			);
+		}
+
+		setIsBusy( false );
+	}
+
+	return (
+		<Flex as="form" onSubmit={ createFromWidgets }>
+			<FlexBlock>
+				<SelectControl
+					label={ __( 'Import widget area' ) }
+					value={ selectedSidebar }
+					options={ options }
+					onChange={ ( value ) => setSelectedSidebar( value ) }
+					disabled={ ! options.length }
+					__next36pxDefaultSize
+					__nextHasNoMarginBottom
+				/>
+			</FlexBlock>
+			<FlexItem
+				style={ {
+					marginBottom: '8px',
+					marginTop: 'auto',
```

Comment:
What is this for?

Reply:
This was needed to align select and submit. A similar solution was already used in one of the components, so I thought it was okay to re-use it here.

Reply:
Gotcha. I think something like this is the way things are headed. But I tried it out quickly and it didn't work 😛

```js
<HStack as="form" onSubmit={ createFromWidgets }>
	<SelectControl
		label={ __( 'Import widget area' ) }
		value={ selectedSidebar }
		options={ options }
		onChange={ ( value ) => setSelectedSidebar( value ) }
		disabled={ ! options.length }
		__next36pxDefaultSize
		__nextHasNoMarginBottom
	/>
	<Spacer marginTop="auto" marginBottom={ 2 }>
		<Button
			variant="primary"
			type="submit"
			isBusy={ isBusy }
			aria-disabled={ isBusy || ! selectedSidebar }
		>
			{ __( 'Import' ) }
		</Button>
	</Spacer>
</HStack>
```

Feel free to continue trying or dismiss this.

Reply:
I tried using the `Spacer` but failed to make it work. It was breaking the layout.

P.S. I think we need some guides and an example for new utility components. Currently, I'm constantly checking recently refactored UI to find the best new practices.